File: src/main/java/org/dbtools/query/shared/filter/NullFilter.java
package org.dbtools.query.shared.filter;

import org.dbtools.query.shared.QueryBuilder;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public class NullFilter extends CompareFilter {

    private boolean isNull;

    public static NullFilter create(@Nullable String field) {
        if (field == null) {
            return null;
        }

        NullFilter filterFormatter = new NullFilter();
        filterFormatter.filter = NullFilter.newInstance(field, true);
        return filterFormatter;
    }

    public static NullFilter create(String field, boolean isNull) {
        NullFilter filterFormatter = new NullFilter();
        filterFormatter.filter = NullFilter.newInstance(field, isNull);
        return filterFormatter;
    }

    private static NullFilter newInstance(String field, boolean isNull) {
        return new NullFilter(field, isNull);
    }

    protected NullFilter() {
        super();
    }

    private NullFilter(String field, boolean isNull) {
        super(field, null, null);
        this.isNull = isNull;
    }

    @Override
    public String build(@Nonnull QueryBuilder queryBuilder) {
        StringBuilder builder = new StringBuilder(field);
        if (isNull) {
            builder.append(" IS NULL");
        } else {
            builder.append(" NOT NULL");
        }
        return builder.toString();
    }

    public NullFilter and(String field) {
        and(NullFilter.create(field));
        return this;
    }

    public NullFilter and(String field, boolean isNull) {
        and(NullFilter.create(field, isNull));
        return this;
    }

    public NullFilter or(String field) {
        or(NullFilter.create(field));
        return this;
    }

    public NullFilter or(String field, boolean isNull) {
        or(NullFilter.create(field, isNull));
        return this;
    }

    @Override
    public NullFilter clone() {
        NullFilter clone = (NullFilter) super.clone();
        clone.isNull = this.isNull;
        return clone;
    }
}

File: dbtools-query/src/commonMain/kotlin/org/dbtools/query/shared/filter/NullFilter.kt
package org.dbtools.query.shared.filter

import org.dbtools.query.shared.CompareType
import org.dbtools.query.shared.QueryBuilder

class NullFilter(
    field: String,
    private var isNull: Boolean
) : CompareFilter(field, if (isNull) CompareType.IS_NULL else CompareType.NOT_NULL, null) {

    override fun build(queryBuilder: QueryBuilder): String = buildString {
        append(field)
        if (isNull) {
            append(" IS NULL")
        } else {
            append(" NOT NULL")
        }
    }

    fun and(field: String): NullFilter {
        and(create(field))
        return this
    }

    fun and(field: String, isNull: Boolean): NullFilter {
        and(create(field, isNull))
        return this
    }

    fun or(field: String): NullFilter {
        or(create(field))
        return this
    }

    fun or(field: String, isNull: Boolean): NullFilter {
        or(create(field, isNull))
        return this
    }

    companion object {
        fun create(field: String): NullFilter {
            val filterFormatter = NullFilter(field, true)
            filterFormatter.filter = newInstance(field, true)
            return filterFormatter
        }

        fun create(field: String, isNull: Boolean): NullFilter {
            val filterFormatter = NullFilter(field, isNull)
            filterFormatter.filter = newInstance(field, isNull)
            return filterFormatter
        }

        private fun newInstance(field: String, isNull: Boolean): NullFilter {
            return NullFilter(field, isNull)
        }
    }
}
